src/GoTrueClient.ts

```diff
@@ -380,25 +380,112 @@ export default class GoTrueClient {
     }
   }
 
+  /**
+   * Sets the session from the provided session information. The access_token
+   * is reused if it is not expired, otherwise a new access token is fetched by
+   * refreshing the session with the provided refresh_token.
+   *
+   * This method is useful when using in a server-side rendered context.
+   *
+   * @param params.refresh_token A valid refresh token (typically obtained from a cookie)
+   * @param params.access_token An access token (typically obtained from a cookie)
+   * @param params.provider_token A provider token (from social login, for example), optional
+   * @param params.expires_in A timeout in seconds when the access token will expire, if not provided taken from the `exp` claim in the access token
+   */
+  async setSession(params: {
+    refresh_token: string
+    access_token: string
+    expires_in?: number
+    provider_token?: string
```

Comment:
can we assign this to a type here so we don't have to check the types below (lines 424-429)?

Reply:
Nope, that's unfortunately a limitation with TypeScript and function overloading as described here: https://www.typescriptlang.org/docs/handbook/2/functions.html#function-overloads

Tho it may be possible to do away a lot of the `typeof` statements (but not all) with some type changes on `arg0`.